src/tree/common_row_partitioner.h

```diff
@@ -106,26 +106,39 @@ class CommonRowPartitioner {
 
   template <typename ExpandEntry>
   void FindSplitConditions(const std::vector<ExpandEntry>& nodes, const RegTree& tree,
-                           const GHistIndexMatrix& gmat, std::vector<int32_t>* split_conditions) {
+                           const GHistIndexMatrix& gmat,
+                           std::vector<int32_t>* split_conditions, bool is_index) {
     auto const& ptrs = gmat.cut.Ptrs();
     auto const& vals = gmat.cut.Values();
 
     for (std::size_t i = 0; i < nodes.size(); ++i) {
       bst_node_t const nidx = nodes[i].nid;
       bst_feature_t const fidx = tree.SplitIndex(nidx);
       float const split_pt = tree.SplitCond(nidx);
-      std::uint32_t const lower_bound = ptrs[fidx];
-      std::uint32_t const upper_bound = ptrs[fidx + 1];
-      bst_bin_t split_cond = -1;
-      // convert floating-point split_pt into corresponding bin_id
-      // split_cond = -1 indicates that split_pt is less than all known cut points
-      CHECK_LT(upper_bound, static_cast<uint32_t>(std::numeric_limits<int32_t>::max()));
-      for (auto bound = lower_bound; bound < upper_bound; ++bound) {
-        if (split_pt == vals[bound]) {
-          split_cond = static_cast<bst_bin_t>(bound);
+      if (is_index) {
+        // if the split_pt is already recorded as a bin_id, use it directly
+        (*split_conditions)[i] = static_cast<int32_t>(split_pt);
+        // at this point, each participants received the best split index,
+        // therefore can recover the split_pt from bin_id, update tree info
+        auto split_pt_local = vals[split_pt];
+        // make updates to the tree, replacing the existing index
+        // with cut value, so as to be consistent with the tree model format
+        const_cast<RegTree::Node&>(tree.GetNodes()[nidx]).SetSplit(fidx, split_pt_local);
```

Comment:
Unless there's really good reason (like interfacing legacy C libraries), const_cast is discouraged, please consider making the necessary cleanup/refactor instead of pushing an experimental edit.

Aside, could you please help look into whether it's possible to extract this float -> integer bin index logic, and move the tree split logic into the evaluator? Currently, it's the only place the modifies a tree, and it would be great if we could keep it that way.

Reply:
yeah... I tried to find a way to make updates to the split value, but ended up with this solution, will revisit to see if there are alternative ways, maybe adding a function to tree node spec

Reply:
moved this "split_value" recovery process to evaluate_splits.h, right after the global best split allgather